swarm/src/handler.rs

```diff
@@ -334,98 +331,69 @@ pub enum ProtocolsChange<'a> {
-        }))
-    }
-
     /// Compute the [`ProtocolsChange`]s required to go from `existing_protocols` to `new_protocols`.
-    pub(crate) fn from_full_sets(
-        existing_protocols: &'a HashSet<StreamProtocol>,
-        new_protocols: &'a HashSet<StreamProtocol>,
-    ) -> SmallVec<[Self; 2]> {
-        if existing_protocols == new_protocols {
-            return SmallVec::new();
-        }
-
-        let mut changes = SmallVec::new();
-
-        let mut added_protocols = new_protocols.difference(existing_protocols).peekable();
-        let mut removed_protocols = existing_protocols.difference(new_protocols).peekable();
-
-        if added_protocols.peek().is_some() {
-            changes.push(ProtocolsChange::Added(ProtocolsAdded {
-                protocols: added_protocols,
-            }));
+    pub(crate) fn from_full_sets<T: AsRef<str>>(
+        existing_protocols: &mut HashMap<AsStrHashEq<T>, bool>,
+        new_protocols: impl IntoIterator<Item = T>,
+        temp_owner: &'a mut Vec<StreamProtocol>,
+    ) -> impl Iterator<Item = Self> {
+        temp_owner.clear();
+        // to avoid looping trough the map to just set the booleans we use the fact that all of
+        // the booleans in the map have same value
+        let valid_value = !existing_protocols.values().next().copied().unwrap_or(false);
+        let mut count = Some(existing_protocols.len());
+        for new_protocol in new_protocols {
+            *existing_protocols
+                .entry(AsStrHashEq(new_protocol))
+                .or_insert_with_key(|k| {
+                    temp_owner.extend(StreamProtocol::try_from_owned(k.0.as_ref().to_owned()).ok());
+                    count = None;
+                    false
+                }) = valid_value;
+            count = count.map(|c| c - 1);
         }
 
-        if removed_protocols.peek().is_some() {
-            changes.push(ProtocolsChange::Removed(ProtocolsRemoved {
-                protocols: Either::Left(removed_protocols),
-            }));
+        if count == Some(0) {
+            return None.into_iter().chain(None);
         }
 
-        changes
+        let added_count = temp_owner.len();
+        existing_protocols.retain(|k, v| {
+            if *v != valid_value {
+                temp_owner.push(StreamProtocol::try_from_owned(k.0.as_ref().to_owned()).unwrap());
+            }
+            *v == valid_value
+        });
+
+        let (added, removed) = temp_owner.split_at(added_count);
+        added
+            .is_empty()
+            .not()
+            .then_some(ProtocolsChange::Added(ProtocolsAdded {
+                protocols: added.iter(),
+            }))
+            .into_iter()
+            .chain(
+                removed
+                    .is_empty()
+                    .not()
+                    .then_some(ProtocolsChange::Removed(ProtocolsRemoved {
+                        protocols: removed.iter(),
+                    })),
+            )
```

Comment:
I find this hard to read. Why did we have to change the API away from returning a `SmallVec`?

Reply:
thats because the Protocols added implements Iterator<Item = &'a StreamProtocol>`, If I own the SmallVec, its impossible to do that.

Reply:
Or right, sorry I got it confused, will do.